File: src/main/java/com/chimbori/crux/MetadataHelpers.java
package com.chimbori.crux;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class MetadataHelpers {
  private MetadataHelpers() {
  }

  private static final Set<String> IGNORED_TITLE_PARTS = new HashSet<>(Arrays.asList(
      "hacker news", "facebook"
  ));

  static String extractTitle(Document doc) {
    try {
      return cleanTitle(new HeuristicString(doc.title())
          .or(StringUtils.innerTrim(doc.select("head title").text()))
          .or(StringUtils.innerTrim(doc.select("head meta[name=title]").attr("content")))
          .or(StringUtils.innerTrim(doc.select("head meta[property=og:title]").attr("content")))
          .or(StringUtils.innerTrim(doc.select("head meta[name=twitter:title]").attr("content")))
          .toString());
    } catch (HeuristicString.CandidateFound candidateFound) {
      return cleanTitle(candidateFound.candidate);
    }
  }

  static String extractAmpUrl(Document doc) {
    try {
      return new HeuristicString(StringUtils.urlEncodeSpaceCharacter(doc.select("link[rel=amphtml]").attr("href")))
          .toString();
    } catch (HeuristicString.CandidateFound candidateFound) {
      return candidateFound.candidate;
    }
  }

  static String extractCanonicalUrl(Document doc) {
    try {
      return new HeuristicString(null)
          .or(StringUtils.urlEncodeSpaceCharacter(doc.select("head link[rel=canonical]").attr("href")))
          .or(StringUtils.urlEncodeSpaceCharacter(doc.select("head meta[property=og:url]").attr("content")))
          .or(StringUtils.urlEncodeSpaceCharacter(doc.select("head meta[name=twitter:url]").attr("content")))
          .toString();
    } catch (HeuristicString.CandidateFound candidateFound) {
      return candidateFound.candidate;
    }
  }

  static String extractDescription(Document doc) {
    try {
      return new HeuristicString(null)
          .or(StringUtils.innerTrim(doc.select("head meta[name=description]").attr("content")))
          .or(StringUtils.innerTrim(doc.select("head meta[property=og:description]").attr("content")))
          .or(StringUtils.innerTrim(doc.select("head meta[name=twitter:description]").attr("content")))
          .toString();
    } catch (HeuristicString.CandidateFound candidateFound) {
      return candidateFound.candidate;
    }
  }

  static String extractSiteName(Document doc) {
    try {
      return new HeuristicString(null)
          .or(StringUtils.innerTrim(doc.select("head meta[property=og:site_name]").attr("content")))
          .or(StringUtils.innerTrim(doc.select("head meta[name=application-name]").attr("content")))
          .toString();
    } catch (HeuristicString.CandidateFound candidateFound) {
      return candidateFound.candidate;
    }
  }

  static String extractThemeColor(Document doc) {
    return doc.select("meta[name=theme-color]").attr("content");
  }

  static String extractImageUrl(Document doc, List<Article.Image> images) {
    try {
      return new HeuristicString(null)
          // Twitter Cards and Open Graph images are usually higher quality, so rank them first.
          .or(StringUtils.urlEncodeSpaceCharacter(doc.select("head meta[name=twitter:image]").attr("content")))
          .or(StringUtils.urlEncodeSpaceCharacter(doc.select("head meta[property=og:image]").attr("content")))
          // Then, grab any hero images from the article itself.
          .or(images != null && images.size() > 0 ? StringUtils.urlEncodeSpaceCharacter(images.get(0).src) : null)
          // image_src or thumbnails are usually low quality, so prioritize them *after* article images.
          .or(StringUtils.urlEncodeSpaceCharacter(doc.select("link[rel=image_src]").attr("href")))
          .or(StringUtils.urlEncodeSpaceCharacter(doc.select("head meta[name=thumbnail]").attr("content")))
          .toString();
    } catch (HeuristicString.CandidateFound candidateFound) {
      return candidateFound.candidate;
    }
  }

  static String extractFeedUrl(Document doc) {
    try {
      return new HeuristicString(null)
          .or(doc.select("link[rel=alternate]").select("link[type=application/rss+xml]").attr("href"))
          .or(doc.select("link[rel=alternate]").select("link[type=application/atom+xml]").attr("href"))
          .toString();
    } catch (HeuristicString.CandidateFound candidateFound) {
      return candidateFound.candidate;
    }
  }

  static String extractVideoUrl(Document doc) {
    return StringUtils.urlEncodeSpaceCharacter(doc.select("head meta[property=og:video]").attr("content"));
  }

  static String extractFaviconUrl(Document doc) {
    try {
      return new HeuristicString(null)
          .or(StringUtils.urlEncodeSpaceCharacter(doc.select("head link[rel=icon]").attr("href")))
          .or(StringUtils.urlEncodeSpaceCharacter(doc.select("head link[rel^=shortcut],link[rel$=icon]").attr("href")))
          .toString();
    } catch (HeuristicString.CandidateFound candidateFound) {
      return candidateFound.candidate;
    }
  }

  static Collection<String> extractKeywords(Document doc) {
    String content = StringUtils.innerTrim(doc.select("head meta[name=keywords]").attr("content"));

    if (content.startsWith("[") && content.endsWith("]")) {
      content = content.substring(1, content.length() - 1);
    }

    String[] split = content.split("\\s*,\\s*");
    if (split.length > 1 || (split.length > 0 && !"".equals(split[0]))) {
      return Arrays.asList(split);
    }
    return Collections.emptyList();
  }

  /**
   * Extracts a set of images from the article content itself. This extraction must be run before
   * the postprocess step, because that step removes tags that are useful for image extraction.
   */
  static List<Article.Image> extractImages(Element topNode) {
    List<Article.Image> images = new ArrayList<>();
    if (topNode == null) {
      return images;
    }

    Elements imgElements = topNode.select("img");
    if (imgElements.isEmpty() && topNode.parent() != null) {
      imgElements = topNode.parent().select("img");
    }

    int maxWeight = 0;
    double score = 1;
    for (Element imgElement : imgElements) {
      Article.Image image = Article.Image.from(imgElement);
      if (image.src.isEmpty() || StringUtils.isAdImage(image.src)) {
        continue;
      }

      image.weight += image.height >= 50 ? 20 : -20;
      image.weight += image.width >= 50 ? 20 : -20;
      image.weight += image.src.startsWith("data:") ? -50 : 0;
      image.weight += image.src.endsWith(".gif") ? -20 : 0;
      image.weight += image.src.endsWith(".jpg") ? 5 : 0;
      image.weight += image.alt.length() > 35 ? 20 : 0;
      image.weight += image.title.length() > 35 ? 20 : 0;
      image.weight += image.noFollow ? -40 : 0;

      image.weight = (int) (image.weight * score);
      if (image.weight > maxWeight) {
        maxWeight = image.weight;
        score = score / 2;
      }

      images.add(image);
    }

    Collections.sort(images, new ImageWeightComparator());
    Log.i("images: %s", images);
    return images;
  }

  static String cleanTitle(String title) {
    StringBuilder res = new StringBuilder();
//        int index = title.lastIndexOf("|");
//        if (index > 0 && title.length() / 2 < index)
//            title = title.substring(0, index + 1);

    int counter = 0;
    String[] strs = title.split("\\|");
    for (String part : strs) {
      if (IGNORED_TITLE_PARTS.contains(part.toLowerCase().trim())) {
        continue;
      }
      if (counter == strs.length - 1 && res.length() > part.length()) {
        continue;
      }
      if (counter > 0) {
        res.append("|");
      }
      res.append(part);
      counter++;
    }
    return StringUtils.innerTrim(res.toString());
  }

  /**
   * Returns the highest-scored image first.
   */
  private static class ImageWeightComparator implements Comparator<Article.Image> {
    @Override
    public int compare(Article.Image o1, Article.Image o2) {
      return o2.weight - o1.weight;
    }
  }
}

File: src/main/kotlin/com/chimbori/crux/extractors/MetadataHelpers.kt
package com.chimbori.crux.extractors

import com.chimbori.crux.common.cleanTitle
import com.chimbori.crux.common.nullIfBlank
import com.chimbori.crux.common.removeWhiteSpace
import java.util.Locale
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element

public fun Document.extractTitle(): String? = (
    title().nullIfBlank()
      ?: select("title").text().nullIfBlank()
      ?: select("meta[name=title]").attr("content").nullIfBlank()
      ?: select("meta[property=og:title]").attr("content").nullIfBlank()
      ?: select("meta[name=twitter:title]").attr("content").nullIfBlank()
    )?.cleanTitle()?.nullIfBlank()

public fun Document.extractCanonicalUrl(): String? = (
    select("link[rel=canonical]").attr("abs:href").nullIfBlank()
      ?: select("meta[property=og:url]").attr("content").nullIfBlank()
      ?: select("meta[name=twitter:url]").attr("content").nullIfBlank()
    )?.removeWhiteSpace()?.nullIfBlank()

public fun Document.extractPaginationUrl(baseUrl: HttpUrl?, nextOrPrev: String): HttpUrl? = (
    select("link[rel=$nextOrPrev]").attr("abs:href").nullIfBlank()
    )?.removeWhiteSpace()?.nullIfBlank()
  ?.let { relativeUrl -> baseUrl?.resolve(relativeUrl) ?: relativeUrl.toHttpUrlOrNull() }

public fun Document.extractDescription(): String? = (
    select("meta[name=description]").attr("content").nullIfBlank()
      ?: select("meta[property=og:description]").attr("content").nullIfBlank()
      ?: select("meta[name=twitter:description]").attr("content").nullIfBlank()
    )?.removeWhiteSpace()?.nullIfBlank()

public fun Document.extractSiteName(): String? = (
    select("meta[property=og:site_name]").attr("content").nullIfBlank()
      ?: select("meta[name=application-name]").attr("content").nullIfBlank()
    )?.removeWhiteSpace()?.nullIfBlank()

public fun Document.extractThemeColor(): String? =
  select("meta[name=theme-color]").attr("content").nullIfBlank()

public fun Document.extractCreated(): String? = (
    select("meta[itemprop=dateCreated]").attr("content").nullIfBlank()
      ?: select("meta[property=article:published_time]").attr("content").nullIfBlank()
      ?: select("meta[property=article:published]").attr("content").nullIfBlank()
    )?.removeWhiteSpace()?.nullIfBlank()

public fun Document.extractModified(): String? = (
    select("meta[itemprop=dateModified]").attr("content").nullIfBlank()
      ?: select("meta[property=article:modified_time]").attr("content").nullIfBlank()
      ?: select("meta[property=article:modified]").attr("content").nullIfBlank()
    )?.removeWhiteSpace()?.nullIfBlank()

public fun Document.extractKeywords(): List<String> =
  select("meta[name=keywords]").attr("content")
    .removeWhiteSpace()
    .removePrefix("[")
    .removeSuffix("]")
    .split("\\s*,\\s*".toRegex())
    .filter { it.isNotBlank() }

public fun Document.extractFaviconUrl(baseUrl: HttpUrl?): HttpUrl? {
  val allPossibleIconElements = listOf(
    select("link[rel~=apple-touch-icon]"),
    select("link[rel~=apple-touch-icon-precomposed]"),
    select("link[rel~=icon]"),
    select("link[rel~=ICON]"),
  )
  return findLargestIcon(allPossibleIconElements.flatten())
    ?.let { baseUrl?.resolve(it) ?: it.toHttpUrlOrNull() }
    ?: baseUrl?.newBuilder()?.encodedPath("/favicon.ico")?.build()
}

public fun Document.extractImageUrl(baseUrl: HttpUrl?): HttpUrl? = (
    // Twitter Cards and Open Graph images are usually higher quality, so rank them first.
    select("meta[name=twitter:image]").attr("content").nullIfBlank()
      ?: select("meta[property=og:image]").attr("content").nullIfBlank()
      // image_src or thumbnails are usually low quality, so prioritize them *after* article images.
      ?: select("link[rel=image_src]").attr("href").nullIfBlank()
      ?: select("meta[name=thumbnail]").attr("content").nullIfBlank()
    )?.let { baseUrl?.resolve(it) ?: it.toHttpUrlOrNull() }

public fun Document.extractFeedUrl(baseUrl: HttpUrl?): HttpUrl? = (
    select("link[rel=alternate]").select("link[type=application/rss+xml]").attr("href").nullIfBlank()
      ?: select("link[rel=alternate]").select("link[type=application/atom+xml]").attr("href").nullIfBlank()
    )?.let { baseUrl?.resolve(it) ?: it.toHttpUrlOrNull() }

public fun Document.extractAmpUrl(baseUrl: HttpUrl?): HttpUrl? =
  select("link[rel=amphtml]").attr("href").nullIfBlank()
    ?.let { baseUrl?.resolve(it) ?: it.toHttpUrlOrNull() }

public fun Document.extractVideoUrl(baseUrl: HttpUrl?): HttpUrl? =
  select("meta[property=og:video]").attr("content").nullIfBlank()
    ?.let { baseUrl?.resolve(it) ?: it.toHttpUrlOrNull() }

internal fun findLargestIcon(iconElements: List<Element>): String? =
  iconElements.maxByOrNull { parseSize(it.attr("sizes")) }?.attr("abs:href")?.nullIfBlank()

/**
 * Given a size represented by "WidthxHeight" or "WidthxHeight ...", will return the largest dimension found.
 *
 * Examples: "128x128" will return 128.
 * "128x64" will return 64.
 * "24x24 48x48" will return 48.
 *
 * @param sizes String representing the sizes.
 * @return largest dimension, or 0 if input could not be parsed.
 */
internal fun parseSize(sizeString: String?): Int {
  if (sizeString.isNullOrBlank()) return 0

  val sizes = sizeString.trim(' ').lowercase(Locale.getDefault())
  return when {
    // For multiple sizes in the same String, split and parse recursively.
    sizes.contains(" ") -> sizes.split(" ").maxOfOrNull { parseSize(it) } ?: 0
    // For handling sizes of format 128x128 etc.
    sizes.contains("x") -> try {
      sizes.split("x").maxOf { it.trim().toInt() }
    } catch (e: NumberFormatException) {
      0
    }
    else -> 0
  }
}
